File: packages/design-ir/src/diff.ts
import { canonicalStringify } from "./canonical";
import type { DesignDocument, DesignNode, SemanticDiff } from "./types";

const GEOMETRY_KEYS = new Set(["transform", "bounds", "x", "y", "width", "height", "rotation_deg"]);
const TEXT_KEYS = new Set(["content", "spans"]);
const ASSET_KEYS = new Set(["asset_id", "source_artifact_version_id"]);
const CONSTRAINT_KEYS = new Set(["constraint_refs"]);

function changedKeys(before: DesignNode, after: DesignNode): readonly string[] {
  const keys = new Set([...Object.keys(before), ...Object.keys(after)]);
  return [...keys].filter(
    (key) =>
      canonicalStringify(before[key] ?? null) !== canonicalStringify(after[key] ?? null),
  );
}

export function computeSemanticDiff(
  before: DesignDocument,
  after: DesignDocument,
): SemanticDiff {
  const beforeIds = new Set(Object.keys(before.nodes));
  const afterIds = new Set(Object.keys(after.nodes));
  const common = [...beforeIds].filter((id) => afterIds.has(id));
  const properties = new Set<string>();
  const text = new Set<string>();
  const geometry = new Set<string>();
  const assets = new Set<string>();
  const constraints = new Set<string>();

  for (const id of common) {
    const left = before.nodes[id]!;
    const right = after.nodes[id]!;
    for (const key of changedKeys(left, right)) {
      if (TEXT_KEYS.has(key)) text.add(id);
      else if (GEOMETRY_KEYS.has(key)) geometry.add(id);
      else if (ASSET_KEYS.has(key)) assets.add(id);
      else if (CONSTRAINT_KEYS.has(key)) constraints.add(id);
      else properties.add(`${id}:${key}`);
    }
  }
  return {
    nodes_added: [...afterIds].filter((id) => !beforeIds.has(id)).sort(),
    nodes_removed: [...beforeIds].filter((id) => !afterIds.has(id)).sort(),
    properties_changed: [...properties].sort(),
    text_changed: [...text].sort(),
    geometry_changed: [...geometry].sort(),
    asset_replaced: [...assets].sort(),
    constraints_changed: [...constraints].sort(),
  };
}

File: packages-py/design_ir/diff.py
from __future__ import annotations

from .canonical import canonical_stringify
from .models import DesignDocument

GEOMETRY_KEYS = {"transform", "bounds", "x", "y", "width", "height", "rotation_deg"}
TEXT_KEYS = {"content", "spans"}
ASSET_KEYS = {"asset_id", "source_artifact_version_id"}
CONSTRAINT_KEYS = {"constraint_refs"}


def compute_semantic_diff(before: DesignDocument, after: DesignDocument) -> dict[str, list[str]]:
    before_nodes = before["nodes"]
    after_nodes = after["nodes"]
    before_ids = set(before_nodes)
    after_ids = set(after_nodes)
    properties: set[str] = set()
    text: set[str] = set()
    geometry: set[str] = set()
    assets: set[str] = set()
    constraints: set[str] = set()

    for node_id in before_ids & after_ids:
        left = before_nodes[node_id]
        right = after_nodes[node_id]
        for key in set(left) | set(right):
            if canonical_stringify(left.get(key)) == canonical_stringify(right.get(key)):
                continue
            if key in TEXT_KEYS:
                text.add(node_id)
            elif key in GEOMETRY_KEYS:
                geometry.add(node_id)
            elif key in ASSET_KEYS:
                assets.add(node_id)
            elif key in CONSTRAINT_KEYS:
                constraints.add(node_id)
            else:
                properties.add(f"{node_id}:{key}")
    return {
        "nodes_added": sorted(after_ids - before_ids),
        "nodes_removed": sorted(before_ids - after_ids),
        "properties_changed": sorted(properties),
        "text_changed": sorted(text),
        "geometry_changed": sorted(geometry),
        "asset_replaced": sorted(assets),
        "constraints_changed": sorted(constraints),
    }
